File: Leetcode/Python/Q2_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        This function takes two linked lists, l1 and l2, which represent two non-negative integers.
        The digits are stored in reverse order, and each of their nodes contains a single digit.
        The function returns the sum as a linked list.
        
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """

        # Create a dummy node that will serve as the starting point for the result linked list.
        dummy = ListNode(0)

        # 'curr' will point to the last node in the result linked list.
        curr = dummy

        # Initialize carry to 0, as there's no carry initially.
        carry = 0

        # Loop until both linked lists are fully traversed.
        while(l1 or l2):
            # Get the value of the current node in l1, or 0 if l1 is null.
            x = l1.val if(l1 != None) else 0
            # Get the value of the current node in l2, or 0 if l2 is null.
            y = l2.val if(l2 != None) else 0

            # Calculate the total sum of the two digits and the carry.
            total = x + y + carry

            # Update carry for the next addition (1 if total >= 10, else 0).
            carry = total // 10

            # Create a new node with the digit value of (total % 10) and attach it to the result list.
            curr.next = ListNode(total % 10)

            # Move the current pointer to the next node in the result list.
            curr = curr.next

            # Move to the next nodes in l1 and l2, if available.
            if(l1): 
                l1 = l1.next
            if(l2): 
                l2 = l2.next

        # If there's a carry left after the last addition, add a new node with the carry value.
        if(carry > 0):
            curr.next = ListNode(carry)

        # Return the next node of dummy, which is the head of the resulting linked list.
        return dummy.next

File: Leetcode/Java/Q2_Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {

    /**
     * This method adds two numbers represented by two linked lists.
     * The digits are stored in reverse order in the linked lists, 
     * so the least significant digit comes first.
     * The method returns the sum as a new linked list.
     *
     * @param head1 - The head of the first linked list.
     * @param head2 - The head of the second linked list.
     * @return The head of the resulting linked list representing the sum.
     */
    public ListNode addTwoNumbers(ListNode head1, ListNode head2) {
        // Create a dummy node to serve as the starting point for the result list.
        ListNode dummy = new ListNode(0);

        // Pointers to traverse the input linked lists and build the result list.
        ListNode p = head1, q = head2, curr = dummy;

        // Initialize carry to handle sums greater than 9.
        int carry = 0;

        // Traverse both linked lists until all nodes are processed.
        while (p != null || q != null) {
            // Get the values from the current nodes of both lists (0 if the node is null).
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;

            // Calculate the sum of the current digits and the carry.
            int sum = x + y + carry;

            // Update carry for the next iteration (carry can only be 0 or 1).
            carry = sum / 10;

            // Create a new node for the current digit of the sum and add it to the result list.
            curr.next = new ListNode(sum % 10);

            // Move the current pointer forward.
            curr = curr.next;

            // Move to the next node in each list if available.
            if (p != null) {
                p = p.next;
            }
            if (q != null) {
                q = q.next;
            }
        }

        // If there's still a carry after the final iteration, add a new node for it.
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }

        // Return the head of the resulting linked list, which is next to the dummy node.
        return dummy.next;
    }
}
